gobblin-data-management/src/main/java/gobblin/data/management/partition/FileSet.java

```diff
@@ -13,38 +13,55 @@
 package gobblin.data.management.partition;
 
 import lombok.AccessLevel;
+import lombok.Builder;
 import lombok.Data;
+import lombok.Getter;
 import lombok.NonNull;
 import lombok.RequiredArgsConstructor;
+import lombok.Setter;
+import lombok.Singular;
 
+import java.io.IOException;
 import java.util.Collection;
 import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
 
+import com.beust.jcommander.internal.Maps;
+import com.google.common.base.Preconditions;
 import com.google.common.collect.ImmutableList;
 import com.google.common.collect.Lists;
 
 import gobblin.data.management.copy.CopyEntity;
+import gobblin.data.management.copy.CopyableFile;
 import gobblin.dataset.Dataset;
+import gobblin.util.request_allocation.Request;
+import gobblin.util.request_allocation.Requestor;
+
 
 /**
- * A named subset of {@link File}s in a {@link Dataset}. (Useful for partitions, versions, etc.)
+ * A named subset of {@link File}s in a {@link Dataset}. (Useful for partitions, versions, etc.).
+ *
+ * The actual list of files in this {@link FileSet} is, in ideal circumstances, generated lazily. As such, the method
+ * {@link #getFiles()} should only be called when the actual list of files is needed.
  */
-@Data
-@RequiredArgsConstructor(access = AccessLevel.PRIVATE)
-public class FileSet<T extends CopyEntity> {
+@RequiredArgsConstructor(access = AccessLevel.PROTECTED)
+public abstract class FileSet<T extends CopyEntity> implements Request<FileSet<CopyEntity>> {
```

Comment:
I am concerned about the "abstract" part. Why not accept Iterable<T> in the constructor? An ArrayList implements Iterable so this covers the materialized list part. Someone can provide a materialization in the iterator() method which will take care of the "lazy" part. It think this is a much less intrusive change than making the class "abstract".